lib/id3c/cli/command/etl/fhir.py

```diff
@@ -0,0 +1,712 @@
+            f"Speciment with unexpected «{specimen_identifier.set_name}» barcode «{barcode}»"
+
+        sample = process_sample(db, specimen_identifier.uuid)
+        process_presence_absence_tests(db, resource, sample.id, barcode)
+
+
+def resource_and_resource_type(entry: BundleEntry) -> Tuple[DomainResource, str]:
+    """ Returns the Resource and Resource type of a given Bundle *entry*. """
+    return entry.resource, entry.resource.resource_type
+
+
+def extract_resources(bundle: Bundle) -> Dict[str, List[DomainResource]]:
+    """
+    Returns all top-level FHIR Resources in a given *bundle*, organized by
+    resource type.
+    """
+    resources: Dict[str, List[DomainResource]] = defaultdict(list)
+
+    for entry in bundle.entry:
+        resource, resource_type = resource_and_resource_type(entry)
+        resources[resource_type].append(resource)
+
+    return resources
+
+
+def extract_related_resources(bundle: Bundle, reference_entry: BundleEntry) -> Dict[str, List[DomainResource]]:
+    """
+    Finds all top-level FHIR Resources in a given *bundle* that contain a full
+    URL reference to a *reference_entry*. Returns these Resources organized by
+    resource type.
+    """
+    def is_related_resource(resource: DomainResource, type: str, url: str) -> bool:
+        """
+        Returns True if the given *resource* has a *type* reference equal to the
+        given *url*. Otherwise, returns False.
+        """
+        reference_map = {
+            'Encounter': 'encounter',
+            'Patient': 'subject',
+        }
+
+        reference = getattr(resource, reference_map[type], None)
+        if reference:
+            return reference.reference == url
+
+        return False
+
+    resources: Dict[str, List[DomainResource]] = defaultdict(list)
+
+    reference_type = reference_entry.resource.resource_type
+    reference_url = reference_entry.fullUrl
+
+    for entry in bundle.entry:
+        resource, resource_type = resource_and_resource_type(entry)
+
+        if is_related_resource(resource, reference_type, reference_url):
+            resources[resource_type].append(resource)
+
+    return resources
+
+
+def extract_contained_resources(resource: DomainResource) -> Dict[str, List[DomainResource]]:
+    """
+    Returns the contained Resources in a given *resource* organized by Resoure type.
+    Returns an empty dict if the given *resource* has no contained Resources.
+    """
+    resources: Dict[str, List[DomainResource]] = defaultdict(list)
+
+    if not resource.contained:
+        return resources
+
+    for contained in resource.contained:
+        resources[contained.resource_type].append(contained)
+
+    return resources
+
+
+def assert_required_resource_types_present(resources: Dict[str, List[DomainResource]]):
+    """
+    Raises an :class:`AssertionError` if the given *resources* do not meet the
+    following requirements:
+    * At least one Specimen Resource exists in the given *resources*
```

Comment:
This doesn't work with the Bundle examples from Ellume. Do we want to loosen our expectations for bundles or ask Ellume to include a Specimen Resource?

Reply:
I would be nice to support processing of resources related to Specimens that only refer to them by identifier without including a stub Specimen resource.

Reply:
This seems like a blocker for deploy tomorrow.